### ISSUE TITLE
How to perform efficient element-wise multiplication on sparse matrices

### ISSUE THREAD

**Issue by szarnyasg** [optimatika/ojAlgo#134]
I have been looking for an efficient Java library for sparse matrix operations for a while now (see my [Software Recommendations StackExchange question](https://softwarerecs.stackexchange.com/questions/51330/sparse-matrix-library-for-java) from July). So far, ojAlgo seems to be the best fit for our problems as it can store sparse matrices in little memory and multiply them efficiently. However, we were surprised to find that its _element-wise multiplication_ is significantly slower than its matrix-matrix multiplication operation.

Here is a snippet to reproduce the issue on relatively small, 50k × 50k matrices, consisting only of zeroes. We made sure to consume all nnz elements of the results by iterating on result of the `nonzero()` method.

```java
    @Test
    public void test() {
        int n = 50_000;
        SparseStore<Double> a = SparseStore.PRIMITIVE.make(n, n);
        SparseStore<Double> b = SparseStore.PRIMITIVE.make(n, n);

        System.out.println("A * B");
        SparseStore<Double> ab1 = SparseStore.PRIMITIVE.make(n, n);
        a.multiply(b).get().supplyTo(ab1);
        final ElementView2D<Double, ?> nnz1 = ab1.nonzeros();
        while (nnz1.hasNext()) { nnz1.next(); }

        System.out.println("A .* B");
        SparseStore<Double> ab2 = SparseStore.PRIMITIVE.make(n, n);
        a.operateOnMatching(MULTIPLY, b).supplyTo(ab2);
        final ElementView2D<Double, ?> nnz2 = ab2.nonzeros();
        while (nnz2.hasNext()) { nnz2.next(); }
    }
```

Matrix multiplication is near instant, element-wise multiplication takes ~80 seconds on an i7-2760QM laptop, running Ubuntu and Java 8. We also tried the `modifyMatching` approach suggested in issue #103, but it did not improve the performance.

**Comment by apete** [comment]
The problem is that the `operateOnMatching` methods must function for any operator and any implementation of Access1D - dense/sparse/primitive/generic...

I committed some experimental code yesterday in the `SparseStuff` branch that you can test. (I didn't test it.) https://github.com/optimatika/ojAlgo/tree/SparseStuff

Checked your post on "Software Recommendations": Very surprised to see that most libraries had problems. ojAlgo has limited support for sparse matrices. I always thought this was an area where ojAlgo couldn't match many of the other libraries. Initially ojAlgo had no support for sparse at all - it's a late addition. Sparseness is not always exploited, and since the API was designed for dense matrices some operations don't make sense for sparse.

**Comment by szarnyasg** [comment]
Thanks! We'll experiment with it (both testing and benchmarking it) and report back.

The lack of support for sparse matrices in the Java ecosystem was also a surprise for me - I expected to go to Maven Central, take a library off the shelf and use it. It turned out this isn't really the case as almost every library is geared (sometimes exclusively) towards dense matrices. Also, only a few libraries are multi-threaded and even less support parallelism on sparse matrices. The output of `htop` hints that ojAlgo is multi-threaded for sparse matrix multiplication. Is that observation correct?

**Comment by apete** [comment]
ojAlgo is designed to use multiple threads when possible and effective. The MatrixStore interface defines several multiplication methods. Depending on which you use the various implementations may end up calling different code.

If you think this issue is solved, then please close it.

**Comment by szarnyasg** [comment]
48203f4f71284f0d49eaf85e9f0f36a53c1f56e0 solved it for us, thanks! Looking forward to see it in a release.

**Comment by apete** [comment]
Not sure if you're still interested, but I made some further improvements to some sparse operations. It's currently in the develop branch and will be part of v47 when that's released.

https://github.com/optimatika/ojAlgo/wiki/Sparse-Matrices

I saw in some ojAlgo code you wrote that you always did a supplyTo(...) at the end. That means you always create an additional copy of the matrix - not very efficient.
